homeassistant/components/history/__init__.py

```diff
@@ -337,16 +400,16 @@ def _sorted_states_to_json(
         domain = split_entity_id(ent_id)[0]
         ent_results = result[ent_id]
         if not minimal_response or domain in NEED_ATTRIBUTE_DOMAINS:
-            ent_results.extend(
-                [
-                    native_state
-                    for native_state in (LazyState(db_state) for db_state in group)
-                    if (
-                        domain != SCRIPT_DOMAIN
-                        or native_state.attributes.get(ATTR_CAN_CANCEL)
-                    )
-                ]
-            )
+            if domain == SCRIPT_DOMAIN:
+                ent_results.extend(
+                    [
+                        native_state
+                        for native_state in (LazyState(db_state) for db_state in group)
+                        if native_state.attributes.get(ATTR_CAN_CANCEL)
+                    ]
+                )
+            else:
+                ent_results.extend([LazyState(db_state) for db_state in group])
```

Comment:
I am not sure if it's faster, but it saves creating a whole list.

```suggestion
                ent_results.extend(LazyState(db_state) for db_state in group)
```